src/gadgets/curves/bn256/ec_pairing.rs

```diff
@@ -0,0 +1,303 @@
+        point: &mut BN256SWProjectivePointTwisted<F>,
+        at: &mut BN256SWProjectivePoint<F>,
+    ) -> Self {
+        // Defining b' - the twist coefficient
+        let params = point.x.c0.params.clone();
+        let mut b_twist = BN256Fq2NNField::from_curve_base(cs, &BN256G2Affine::b_coeff(), &params);
+
+        // c0 <- -2 * Y * Z * y_P
+        let mut c0 = point.y.mul(cs, &mut point.z);
+        let mut c0 = c0.mul_c0(cs, &mut at.y);
+        let mut c0 = c0.double(cs);
+        let c0 = c0.negated(cs);
+
+        // c4 <- 3b' * Z^2 - Y^2
+        let mut z2 = point.z.square(cs);
+        let mut z2 = z2.mul(cs, &mut b_twist);
+        let mut c4 = z2.double(cs);
+        let mut c4 = c4.add(cs, &mut z2);
+        let mut y2 = point.y.square(cs);
+        let c4 = c4.sub(cs, &mut y2);
+
+        // c3 <- 3 * X^2 * x_P
+        let mut x2 = point.x.square(cs);
+        let mut c3 = x2.mul_c0(cs, &mut at.x);
+        let mut c3 = c3.double(cs);
+        let c3 = c3.add(cs, &mut x2);
+
+        Self::new(c0, c3, c4)
+    }
+}
+
+/// Struct for the miller loop evaluation for the BN256 curve.
+/// Here, the Miller loop returns the accumulated f value after the loop
+/// without the final exponentiation.
+pub struct MillerLoopEvaluation<F, CS>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    accumulated_f: BN256Fq12NNField<F>,
+    _marker: std::marker::PhantomData<CS>,
+}
+
+impl<F, CS> MillerLoopEvaluation<F, CS>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    /// This function computes the Miller loop for the BN256 curve, using
+    /// algorithm from _Section 2_ from https://eprint.iacr.org/2016/130.pdf.
+    pub fn evaluate(
+        cs: &mut CS,
+        p: &mut BN256SWProjectivePoint<F>,
+        q: &mut BN256SWProjectivePointTwisted<F>,
+    ) -> Self {
+        // Setting evaluation parameters
+        let params = p.x.params.clone();
+        let mut evaluation = LineFunctionEvaluation::zero(cs, &params);
+
+        let mut f1 = BN256Fq12NNField::one(cs, &params);
+        let mut r = q.clone();
+
+        for u in CURVE_PARAMETER_WNAF {
+            // Doubling step: f1 <- f1^2 * L_{R,R}(P), R <- 2R
+            let mut tan_fn = evaluation.at_tangent(cs, &mut r, p);
+            f1 = f1.square(cs);
+            f1 = Self::mul_f12_by_line_fn(cs, &mut f1, &mut tan_fn);
+            r = r.double(cs);
+
+            // Skip if u is zero
+            if u == 0 {
+                continue;
+            }
+
+            // Addition step: f1 <- f1 * L_{R,Q}(P), R <- R + Q.
+            // If u is negative, negate Q.
+            let mut q = q.clone();
+            if u == -1 {
+                q = q.negated(cs);
+            }
+
+            let mut line_fn = evaluation.at_line(cs, &mut r, &mut q, p);
+            f1 = Self::mul_f12_by_line_fn(cs, &mut f1, &mut line_fn);
+
+            let qx = q.x.clone();
+            let qy = q.y.clone();
+            r = r.add_mixed(cs, &mut (qx, qy));
```

Comment:
would there be any case here where `qz` isn't 1 and so we'd need to reduce to affine first?

Reply:
Since the $Q$ point does not change during the loop, as long as the inputted $Q$ is "normalized" ($Z=1$), I believe we are safe to assume $Z=1$.

However, I added the enforcement on `q.z` to be one at the beginning of the function; thanks for noticing that!